src/app/store/zone/slice.ts

```diff
@@ -1,22 +1,249 @@
+import type { PayloadAction } from "@reduxjs/toolkit";
 import { createSlice } from "@reduxjs/toolkit";
 
+import { ZONE_ACTIONS, ZONE_WEBSOCKET_METHODS } from "./constants";
+import type {
+  CreateParams,
+  DeleteParams,
+  UpdateParams,
+  Zone,
+  ZoneActionNames,
+  ZonePayloadActionWithIdentifier,
+  ZoneGenericActions,
+  ZoneModelActions,
+  ZonePK,
+  ZoneState,
+} from "./types";
 import { ZoneMeta } from "./types";
-import type { CreateParams, UpdateParams, ZoneState } from "./types";
 
-import {
-  generateCommonReducers,
-  genericInitialState,
-} from "app/store/utils/slice";
+import { ACTION_STATUS } from "app/base/constants";
+import type { ActionStatuses, APIError } from "app/base/types";
+
+const { cleanup, create, delete: deleteAction, fetch, update } = ZONE_ACTIONS;
+
+const { error, idle, loading, success } = ACTION_STATUS;
+
+export const initialGenericActions: ZoneGenericActions = {
+  [create]: idle,
+  [fetch]: idle,
+};
+
+export const initialModelActions: ZoneModelActions = {
+  [deleteAction]: {
+    [error]: [],
+    [loading]: [],
+    [success]: [],
+  },
+  [update]: {
+    [error]: [],
+    [loading]: [],
+    [success]: [],
+  },
+};
+
+const addError = (
+  state: ZoneState,
+  action: ZoneActionNames,
+  error: APIError,
+  zonePK: ZonePK | null = null
+) => {
+  state.errors.unshift({ action, error, identifier: zonePK });
+};
+
+const updateGenericAction = (
+  state: ZoneState,
+  actionName: keyof ZoneGenericActions,
+  status: ActionStatuses
+) => {
+  state.genericActions[actionName] = status;
+};
+
+const updateModelAction = (
+  state: ZoneState,
+  actionName: keyof ZoneModelActions,
+  status: ActionStatuses,
+  zonePK: ZonePK
+) => {
+  const statuses = state.modelActions[actionName];
+  statuses[error] = statuses[error].filter((id) => id !== zonePK);
+  statuses[loading] = statuses[loading].filter((id) => id !== zonePK);
+  statuses[success] = statuses[success].filter((id) => id !== zonePK);
+  if (status === error) {
+    statuses[error].push(zonePK);
+  } else if (status === loading) {
+    statuses[loading].push(zonePK);
+  } else {
+    statuses[success].push(zonePK);
+  }
+};
 
 const zoneSlice = createSlice({
   name: ZoneMeta.MODEL,
-  initialState: genericInitialState as ZoneState,
-  reducers: generateCommonReducers<
-    ZoneState,
-    ZoneMeta.PK,
-    CreateParams,
-    UpdateParams
-  >(ZoneMeta.MODEL, ZoneMeta.PK),
+  initialState: {
+    errors: [],
+    genericActions: initialGenericActions,
+    items: [],
+    modelActions: initialModelActions,
+  } as ZoneState,
+  reducers: {
+    [cleanup]: (state, action: PayloadAction<ZoneActionNames[]>) => {
```

Comment:
@huwshimi I've updated the cleanup action as you suggested in the previous PR, so now you pass in a list of actions and it clears errors and statuses for those actions only. I thought about having the default be to clear everything, but maybe that promotes encourages behaviour?

Reply:
Yeah think this less likely to lead to unintentional bugs.